# CDAT Migration Regression Testing Notebook (`.json` metrics)

This notebook is used to perform regression testing between the development and
production versions of a diagnostic set.

## How it works

It compares the relative differences (%) between two sets of `.json` files in two
separate directories, one for the refactored code and the other for the `main` branch.

It will display metrics values with relative differences >= 2%. Relative differences are used instead of absolute differences because:

- Relative differences are in percentages, which shows the scale of the differences.
- Absolute differences are just a raw number that doesn't factor in
  floating point size (e.g., 100.00 vs. 0.0001), which can be misleading.

## How to use

PREREQUISITE: The diagnostic set's metrics stored in `.json` files in two directories
(dev and `main` branches).

1. Make a copy of this notebook under `auxiliary_tools/cdat_regression_testing/<DIR_NAME>`.
2. Run `mamba create -n cdat_regression_test -y -c conda-forge "python<3.12" xarray netcdf4 dask pandas matplotlib-base ipykernel`
3. Run `mamba activate cdat_regression_test`
4. Update `DEV_PATH` and `MAIN_PATH` in the copy of your notebook.
5. Run all cells IN ORDER.
6. Review results for any outstanding differences (>= 2%).
   - Debug these differences (e.g., bug in metrics functions, incorrect variable references, etc.)


## Setup Code


In [1]:
import glob
from typing import List

import pandas as pd

from auxiliary_tools.cdat_regression_testing.utils import (
    get_num_metrics_above_diff_thres,
    get_rel_diffs,
    update_diffs_to_pct,
    highlight_large_diffs,
)

SET_NAME = "area_mean_time_series"
SET_DIR = "662-area-mean-time-series"

DEV_PATH = f"/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/{SET_DIR}/{SET_NAME}/**"
MAIN_PATH = f"/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/{SET_NAME}/**"

DEV_GLOB = sorted(glob.glob(DEV_PATH + "/*.json"))
MAIN_GLOB = sorted(glob.glob(MAIN_PATH + "/*.json"))

if len(DEV_GLOB) == 0 or len(MAIN_GLOB) == 0:
    raise IOError("No files found at DEV_PATH and/or MAIN_PATH.")

if len(DEV_GLOB) != len(MAIN_GLOB):
    raise IOError("Number of files do not match at DEV_PATH and MAIN_PATH.")

In [2]:
def get_metrics(filepaths: List[str]) -> pd.DataFrame:
    """Get the metrics using a glob of `.json` metric files in a directory.

    Parameters
    ----------
    filepaths : List[str]
        The filepaths for metrics `.json` files.

    Returns
    -------
    pd.DataFrame
        The DataFrame containing the metrics for all of the variables in
        the results directory.
    """
    metrics = []

    for filepath in filepaths:
        df = pd.read_json(filepath)

        filename = filepath.split("/")[-1]
        var_key = filename.split("-")[0]
        region = filename.split("-")[-1]

        # Add the variable key to the MultiIndex and update the index
        # before stacking to make the DataFrame easier to parse.
        multiindex = pd.MultiIndex.from_product([[var_key], [region], [*df.index]])
        df = df.set_index(multiindex)
        df.stack()

        metrics.append(df)

    df_final = pd.concat(metrics)

    return df_final

In [3]:
metrics = []

for filepath in DEV_GLOB:
    df = pd.read_json(filepath)

    filename = filepath.split("/")[-1]
    var_key = filename.split("-")[0]

    # Add the variable key to the MultiIndex and update the index
    # before stacking to make the DataFrame easier to parse.
    multiindex = pd.MultiIndex.from_product([[var_key], [filename], [*df.index]])
    df = df.set_index(multiindex)
    df.stack()

    metrics.append(df)

df_final = pd.concat(metrics)

## 1. Get the metrics for the development and `main` branches and their differences.


In [4]:
df_metrics_dev = get_metrics(DEV_GLOB)
df_metrics_main = get_metrics(MAIN_GLOB)
df_metrics_diffs = get_rel_diffs(df_metrics_dev, df_metrics_main)

## 2. Filter differences to those above maximum threshold (2%).

All values below maximum threshold will be labeled as `NaN`.

- **If all cells in a row are NaN (< 2%)**, the entire row is dropped to make the results easier to parse.
- Any remaining NaN cells are below < 2% difference and **should be ignored**.


In [5]:
df_metrics_diffs_thres = df_metrics_diffs[df_metrics_diffs >= 0.02]
df_metrics_diffs_thres = df_metrics_diffs_thres.dropna(
    axis=0, how="all", ignore_index=False
)

## 3. Combine all DataFrames to get the final result.


In [6]:
df_metrics_all = pd.concat(
    [df_metrics_dev.add_suffix("_dev"), df_metrics_main.add_suffix("_main")],
    axis=1,
    join="outer",
)
df_final = df_metrics_diffs_thres.join(df_metrics_all)
df_final = update_diffs_to_pct(df_final, ["e3sm_v2 (0051-0060) DIFF (%)"])

In [7]:
df_final

e3sm_v2 (0051-0060) DIFF (%)  e3sm_v2 (0051-0060)_dev  \
FSNTOA 20N50N.json 0                        4.31%               258.548532   
       50N90N.json 0                        7.84%               131.256191   
       50S20S.json 0                        4.91%               241.397554   
       90S50S.json 0                       14.17%                99.968042   
LHFLX  90S50S.json 0                        2.79%                25.029281   
QFLX   90S50S.json 0                        2.79%                 0.865012   
SHFLX  90S50S.json 0                        6.11%                 6.283542   
                   2                        2.21%                 5.656822   
       land.json   0                        2.58%                36.967921   
SWCF   20N50N.json 0                        3.29%               -44.759103   
       50N90N.json 0                        7.61%               -46.444030   
       50S20S.json 0                        4.03%               -50.648058   
       90S50S.json 0                       10.06%               -39.872953   
       land.json   0                        2.21%               -35.209167   
TREFHT 20N50N.json 0                        3.88%                15.302445   
       50N90N.json 0                       27.07%                -3.814642   
                   3                        2.02%                -5.086180   
                   4                        3.64%                -4.984067   
                   6                        2.84%                -6.251635   
                   8                        2.13%                -5.789678   
       90S50S.json 0                        6.22%                -9.728226   
       land.json   0                        6.20%                 7.628299   

                      e3sm_v2 (0051-0060)_main  
FSNTOA 20N50N.json 0                247.400054  
       50N90N.json 0                120.962021  
       50S20S.json 0                253.252342  
       90S50S.json 0                114.128953  
LHFLX  90S50S.json 0                 24.330116  
QFLX   90S50S.json 0                  0.840849  
SHFLX  90S50S.json 0                  5.899695  
                   2                  5.782035  
       land.json   0                 36.013612  
SWCF   20N50N.json 0                -43.287791  
       50N90N.json 0                -42.907371  
       50S20S.json 0                -52.688797  
       90S50S.json 0                -43.884584  
       land.json   0                -34.429798  
TREFHT 20N50N.json 0                 14.709370  
       50N90N.json 0                 -4.847081  
                   3                 -5.188787  
                   4                 -5.165517  
                   6                 -6.074116  
                   8                 -5.913206  
       90S50S.json 0                 -9.123414  
       land.json   0                  7.155572

## 4. Review variables and metrics above difference threshold.

- <span style="color:red">Red</span> cells are differences >= 2%
- `nan` cells are differences < 2% and **should be ignored**


In [8]:
df_final_adj = df_final.reset_index(names=["var_key", "region", "metric"])
get_num_metrics_above_diff_thres(df_metrics_all, df_final_adj)

* Related variables ['FSNTOA', 'LHFLX', 'QFLX', 'SHFLX', 'SWCF', 'TREFHT']
* Number of metrics above 2% max threshold: 22 / 720


In [9]:
highlight_large_diffs(df_final_adj, ["e3sm_v2 (0051-0060) DIFF (%)"])

## Results

- 698 / 720 metrics are within diff tolerance
- 22 / 720 metrics are not within diff tolerance
- Debug large diffs, specifically
  - 3 FSNTOA 90S50S.json 0 14.17% 99.968042 114.128953
  - 12 SWCF 90S50S.json 0 10.06% -39.872953 -43.884584
  - 15 TREFHT 50N90N.json 0 27.07% -3.814642 -4.847081

### Possible explanation for 22 / 720 metric diffs

The cdms2.open slice flag ("co") adds an additional time coordinate
point when opening up time series datasets.

Here's the logic from `dataset.py`:

```python
 if sub_monthly:
     start_time = "{}-01-01".format(start_year)
     end_time = "{}-01-01".format(str(int(end_year) + 1))
     slice_flag = "co"
 else:
     start_time = "{}-01-15".format(start_year)
     end_time = "{}-12-15".format(end_year)
     slice_flag = "ccb"
```

Related issue: https://github.com/E3SM-Project/e3sm_diags/issues/759

Next steps:

- Merge this PR and address issue #759, then re-test area_mean_time_series branch.
